src/tls/letsencrypt.rs

```diff
@@ -0,0 +1,142 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+use mktemp::Temp;
+use std::fs::File;
+use std::io;
+use std::io::Write;
+use std::os::unix::fs::symlink;
+use std::path::{ Path, PathBuf };
+use std::process::Command;
+use std::sync::mpsc::{ channel, Receiver };
+use std::thread;
+use tls::{ CertificateManager, CertificateRecord };
+
+/// Get a SAN certificate from LetsEncrypt for a given list of names.
+pub fn get_san_cert_for<T>(names: T, certificate_manager: CertificateManager,
+                           box_certificate: CertificateRecord, dns_endpoint: String)
+        -> Receiver<io::Result<()>>
+    where T: Iterator<Item=String>,
+          T: DoubleEndedIterator,
+          T: Clone + Send + 'static {
+
+    let (tx, rx) = channel();
+
+    thread::spawn(move || {
+        tx.send(
+            _get_san_cert_for(
+                names, certificate_manager,
+                &box_certificate, &dns_endpoint)
+          ).unwrap();
+    });
+
+    rx
+}
+
+/// Blocking version of get_san_cert_for
+fn _get_san_cert_for<T>(names: T, certificate_manager: CertificateManager,
+                        box_certificate: &CertificateRecord, dns_endpoint: &str)
+    -> io::Result<()>
+    where T: Iterator<Item=String>,
+          T: DoubleEndedIterator,
+          T: Clone + 'static {
+    let temp_dir = try!(Temp::new_dir());
+
+    let mut domains_file = temp_dir.to_path_buf();
+    domains_file.push("domains.txt");
+
+    // Package letsencrypt inside the binary
+    let letsencryptsh = include_str!("scripts/letsencrypt.sh");
+
+    let domains_txt = names.clone().rev().fold("".to_owned(), |accumulator, name| {
+        format!("{} {}", name, accumulator)
+    });
+
+    let mut letsencrypt_file = temp_dir.to_path_buf();
+    letsencrypt_file.push("letsencrypt.sh");
+
+    debug!("Creating letsencrypt client");
+    try!(File::create(letsencrypt_file.clone()).and_then(|mut f| {
+        f.write_all(letsencryptsh.as_bytes())
+    }));
+
+    debug!("Creating domains.txt for letsencrypt client");
+    try!(File::create(domains_file).and_then(|mut f| {
+        f.write_all(domains_txt.as_bytes())
+    }));
```

Comment:
uber-nit: order the lines above so domains.txt related code is in one block and letsencrypt.sh code is in another block